lib/graphql/schema/warden.rb

```diff
@@ -126,8 +126,53 @@ def root_type_for_operation(op_name)
 
       private
 
+      def union_memberships(obj_type)
+        @unions ||=  read_through { |obj_type| @schema.union_memberships(obj_type) }
+        @unions[obj_type]
+      end
+
       def visible_field?(field_defn)
-        visible?(field_defn) && visible?(field_defn.type.unwrap)
+        visible?(field_defn) && visible_type?(field_defn.type.unwrap)
+      end
+
+      def visible_type?(type_defn)
+        return false unless visible?(type_defn)
+
+        if type_defn.kind.object? && has_abstract_type?(type_defn)
+          visible_implementation?(type_defn)
+        else
+          referenced_by_visible_members?(type_defn)
+        end
+      end
+
+      def visible_implementation?(type_defn)
+        if orphan?(type_defn)
+          member_or_implements?(type_defn)
+        else
+          referenced_by_visible_members?(type_defn) || member_or_implements?(type_defn)
+        end
+      end
+
+      def referenced_by_visible_members?(type_defn)
+        members = @schema.get_members_of_type(type_defn.unwrap.name)
+        # Types should be visible if there are
+        # no member definitions with that type
+        return true unless members.any?
+        # Only hide the type when there are fields
+        # but none are visible.
+        members.any? { |member| visible?(member) }
+      end
+
+      def member_or_implements?(type_defn)
+        interfaces(type_defn).any? || union_memberships(type_defn).select { |u| visible?(u) }.any?
```

Comment:
I think `.any? { ... }` is preferable to `.select { ... }.any?` since it will return after finding a first match instead of rebuilding a new array, would that work here?